core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java

```diff
@@ -255,17 +277,159 @@ private void rebuildIndex() throws Exception {
+      PrivateMetadata.Builder builder = new PrivateMetadata.Builder();
+      if (txConfig.transactionMode() == TransactionMode.TRANSACTIONAL && txConfig.lockingMode() == LockingMode.OPTIMISTIC) {
+         builder.entryVersion(new SimpleClusteredVersion(1, 1));
+      }
+
+      // Async XSite
+      if (config.sites().hasAsyncEnabledBackups()) {
+         String siteName = cache.getCacheManager().getTransport().localSiteName();
+         IracEntryVersion version = new IracEntryVersion(Collections.singletonMap(siteName, TopologyIracVersion.newVersion(1)));
+         builder.iracMetadata(new IracMetadata(siteName, version));
+      }
+      return builder.build();
+   }
+
+   private void migrateFromV12_0()  throws Exception {
+      // ISPN-13128 Corrupt migration data can only be created with default marshaller
+      if (ctx.getGlobalConfiguration().serialization().marshaller() == null) {
+         migrateCorruptDataV12_0();
+         return;
+      }
+
+      // Data is not corrupt, so simply update file magic
+      String cacheName = ctx.getCache().getName();
+      PERSISTENCE.startMigratingPersistenceData(cacheName);
+      try {
+         channel.write(ByteBuffer.wrap(MAGIC_LATEST), 0);
+      } catch (IOException e) {
+         throw PERSISTENCE.persistedDataMigrationFailed(cacheName, e);
+      }
+      rebuildIndex();
+   }
+
+   private void migrateCorruptDataV12_0() {
+      String cacheName = ctx.getCache().getName();
+      PERSISTENCE.startRecoveringCorruptPersistenceData(cacheName);
+      File newFile = new File(file.getParentFile(), cacheName + "_new.dat");
+
+      // Day before the eelease of Infinispan 10.0.0.Final
+      // This was the first release that SFS migrations on startup could be migrated from via ISPN 10 -> 11 -> 12
+      long sanityEpoch = LocalDate.of(2019, 10, 26)
+            .atStartOfDay()
+            .atZone(ZoneId.systemDefault())
+            .toInstant()
+            .toEpochMilli();
+      long currentTs = timeService.wallClockTime();
+
+      ByteBuffer buf = ByteBuffer.allocate(KEY_POS_LATEST);
+      try (FileChannel newChannel = SecurityActions.openFileChannel(newFile)) {
+         //Write Magic
+         newChannel.truncate(0);
+         newChannel.write(ByteBuffer.wrap(MAGIC_12_0), 0);
+
+         long oldFilePos = MAGIC_12_0.length;
+         while (true) {
+            buf = readChannel(buf, oldFilePos, KEY_POS_LATEST);
+            // EOF reached
+            if (buf.remaining() > 0)
+               break;
+
+            buf.flip();
+            FileEntry fe = new FileEntry(oldFilePos, buf);
+            // Semantic check to find valid FileEntry
+            if (fe.size <= 0 || fe.keyLen < 0 || fe.dataLen <= 0 || fe.metadataLen < 0 || fe.internalMetadataLen < 0 || fe.expiryTime < -1) {
+               // Check failed, try to read FileEntry from the next byte
+               oldFilePos++;
+               continue;
+            }
+
```

Comment:
```java
If (fe.keyLen == 0) {
  // Just to optimize the previously removed key byte forward search a bit
  oldFilePos += fe.dataLen + fe.metadataLen;
}
```
Since the `unmarshallObject` method below is guaranteed to fail and we know we can skip this amount at least (since fe.size is broken).

Reply:
It's probably unlikely, but isn't it possible that we get a false positive reading `FileEntry` that has `fe.keyLen=0` with other garbage values for `fe.dataLen` etc? The current code is less efficient, but it will guard against this.

Reply:
I think moving to the next byte is the only safe way to handle this, e.g:

```java
If (fe.keyLen == 0) {
  oldFilePos++;
  continue;
}
```

Reply:
Agreed. Forgot that the size could be 0 and the other unmarshalling validations can't be checked.

Reply:
In which case just change if above to be `fe.keyLen <= 0`. Just put a comment that technically 0 is valid, but means we should be skipping the entry.